File: sdk/dotnet/src/Tolap.Core/Enums.cs
namespace Tolap.Core;

/// <summary>
/// Type of masking applied to a field value. Integer values represent restrictiveness order
/// (higher = more restrictive), ranked by how much of the original value is disclosed:
/// <c>partial</c> leaks real characters, <c>hash</c> is irreversible but joinable, <c>full</c>
/// leaks the length, <c>redact</c> leaks nothing, and <c>null</c> leaks not even the field's
/// presence. See canonical-enforcement-spec.md section 6.
/// </summary>
public enum MaskType
{
    Partial = 1,
    Hash = 2,
    Full = 3,
    Redact = 4,
    Null = 5
}

/// <summary>
/// Restrictiveness ranking helpers for <see cref="MaskType"/>.
/// </summary>
public static class MaskTypeExtensions
{
    /// <summary>
    /// Rank of an unrecognized mask type. A typo or a mask type from a newer schema
    /// version must never be beaten by a known-but-weaker type, so it ranks above
    /// every known value.
    /// </summary>
    private const int UnknownRestrictiveness = 6;

    /// <summary>
    /// Ranks a mask type by how little of the value it discloses (higher = stricter).
    /// Anything outside the known set ranks most restrictive so that merging can never
    /// downgrade an unknown mask into a weaker known one.
    /// </summary>
    public static int Restrictiveness(this MaskType maskType) => maskType switch
    {
        MaskType.Partial => 1,
        MaskType.Hash => 2,
        MaskType.Full => 3,
        MaskType.Redact => 4,
        MaskType.Null => 5,
        _ => UnknownRestrictiveness
    };
}

/// <summary>
/// Comparison operator for row filter conditions.
/// </summary>
/// <remarks>
/// Every operator is supported by both enforcement paths: the post-fetch pass in
/// <see cref="EnforcementEngine.ApplyRowFilters"/> and the pre-execution SQL push-down in
/// <see cref="SqlQueryRewriter"/>, except <see cref="Matches"/> whose regex dialect is not
/// portable across engines (see <see cref="SqlQueryRewriter.UnpushableFilters"/>). New
/// members are appended so existing ordinal values are unchanged; the wire form is the
/// camelCase string emitted by <see cref="FilterOperatorJsonConverter"/>, never the ordinal.
/// </remarks>
public enum FilterOperator
{
    Equals,
    NotEquals,
    In,
    NotIn,
    GreaterThan,
    LessThan,
    Contains,
    StartsWith,
    Matches,

    /// <summary>Field value is ordered at or after the comparison value.</summary>
    GreaterThanOrEqual,

    /// <summary>Field value is ordered at or before the comparison value.</summary>
    LessThanOrEqual,

    /// <summary>
    /// Field value matches a SQL <c>LIKE</c> pattern, where <c>%</c> matches any run of
    /// characters and <c>_</c> matches exactly one. Case-sensitive, and <c>\</c> escapes
    /// a literal <c>%</c>, <c>_</c>, or <c>\</c>.
    /// </summary>
    Like,

    /// <summary>Negation of <see cref="Like"/>.</summary>
    NotLike,

    /// <summary>
    /// Field is present on the row and its value is null. A row missing the field
    /// entirely is dropped, per the fail-closed rule in spec section 7.
    /// </summary>
    IsNull,

    /// <summary>Field is present on the row and its value is not null.</summary>
    IsNotNull,

    /// <summary>
    /// Field value falls within an inclusive range given as the first two entries of
    /// <see cref="RowFilter.Values"/>.
    /// </summary>
    Between
}

/// <summary>
/// The kinds of write a policy governs (connector-spec.md section 4.1).
/// </summary>
/// <remarks>
/// <see cref="Insert"/>, <see cref="Update"/> and <see cref="Delete"/> map one-to-one
/// onto <c>canInsert</c>, <c>canUpdate</c> and <c>canDelete</c>. <see cref="Upsert"/> is
/// for a call that cannot be classified as either a create or an overwrite — an
/// unconditional object-store <c>PUT</c>, for example — and requires <b>both</b>
/// <c>canInsert</c> and <c>canUpdate</c>, the safe intersection connector-spec.md
/// section 8 mandates.
/// </remarks>
public enum WriteOperation
{
    /// <summary>Creating a new record or object. Governed by <c>canInsert</c>.</summary>
    Insert,

    /// <summary>Modifying an existing record or object. Governed by <c>canUpdate</c>.</summary>
    Update,

    /// <summary>Removing a record or object. Governed by <c>canDelete</c>.</summary>
    Delete,

    /// <summary>
    /// A write that may create or overwrite and cannot tell which. Requires both
    /// <c>canInsert</c> and <c>canUpdate</c>.
    /// </summary>
    Upsert
}

/// <summary>
/// Type of entity receiving a policy assignment.
/// </summary>
public enum AssigneeType
{
    User,
    Group,
    Role,
    ServiceAccount
}

/// <summary>
/// Cryptographic algorithm used for signing security contexts.
/// </summary>
public enum SigningAlgorithm
{
    HmacSha256,
    HmacSha512,
    Ed25519
}

File: sdk/python/tolap-core/tolap_core/enums.py
from __future__ import annotations

from enum import Enum


class MaskType(Enum):
    full = "full"
    hash = "hash"
    partial = "partial"
    redact = "redact"
    null = "null"

    @property
    def restrictiveness(self) -> int:
        return mask_restrictiveness(self)


# Ranked by how much of the original value is disclosed (canonical spec section 6):
# partial leaks real characters, hash is irreversible but joinable, full leaks the
# length, redact leaks nothing, null leaks not even the field's presence. Higher
# rank wins a merge, so null/redact beat partial rather than losing to it.
_MASK_RESTRICTIVENESS: dict[MaskType, int] = {
    MaskType.partial: 1,
    MaskType.hash: 2,
    MaskType.full: 3,
    MaskType.redact: 4,
    MaskType.null: 5,
}

# An unrecognized mask type (typo, or a newer schema version) must never be beaten
# by a known-but-weaker type, so it ranks above every value above.
_UNKNOWN_MASK_RESTRICTIVENESS = max(_MASK_RESTRICTIVENESS.values()) + 1


def mask_restrictiveness(mask_type: object) -> int:
    """Rank a mask type by how little of the value it discloses (higher = stricter).

    Anything that is not a known MaskType ranks most restrictive so that merging
    can never downgrade an unknown mask into a weaker known one.
    """
    if isinstance(mask_type, MaskType):
        return _MASK_RESTRICTIVENESS[mask_type]
    return _UNKNOWN_MASK_RESTRICTIVENESS


class FilterOperator(Enum):
    """Row-filter comparison operators.

    The member order and the string values are held identical to the ``operator``
    enum in ``schema/v1.0/policy-definition.schema.json`` (and its copy in
    ``effective-policy.schema.json``). A schema-valid policy that this enum cannot
    express is a cross-SDK divergence: the signature still verifies, because the
    canonical payload covers the policy verbatim, so the policy passes every
    integrity check while each SDK produces a different access outcome.
    """

    equals = "equals"
    not_equals = "notEquals"
    in_ = "in"
    not_in = "notIn"
    greater_than = "greaterThan"
    greater_than_or_equal = "greaterThanOrEqual"
    less_than = "lessThan"
    less_than_or_equal = "lessThanOrEqual"
    contains = "contains"
    starts_with = "startsWith"
    like = "like"
    not_like = "notLike"
    matches = "matches"
    is_null = "isNull"
    is_not_null = "isNotNull"
    between = "between"


class WriteOperation(Enum):
    """The kinds of write a policy governs (connector spec section 4.1).

    ``insert``/``update``/``delete`` map one-to-one onto ``canInsert``,
    ``canUpdate`` and ``canDelete``. ``upsert`` is for a call that cannot be
    classified as either a create or an overwrite -- an unconditional object-store
    ``PUT``, for example -- and requires **both** ``canInsert`` and ``canUpdate``,
    the safe intersection connector spec section 8 mandates.
    """

    insert = "insert"
    update = "update"
    delete = "delete"
    upsert = "upsert"


class AssigneeType(Enum):
    user = "user"
    group = "group"
    role = "role"
    service_account = "serviceAccount"


class SigningAlgorithm(Enum):
    hmac_sha256 = "hmac-sha256"
    hmac_sha512 = "hmac-sha512"
    ed25519 = "ed25519"
